source/game/system/KPadController.cc

```diff
@@ -147,6 +147,71 @@ void RaceInputState::reset() {
     trickRaw = 0;
 }
 
+/// @brief Checks if the input state is valid.
+/// @return If the input state is valid.
+bool RaceInputState::isValid() const {
+    if (!isButtonsValid()) {
+        return false;
+    }
+
+    if (!isStickValid(stick.x) || !isStickValid(stick.y)) {
+        return false;
+    }
+
+    if (!isTrickValid()) {
+        return false;
+    }
+
+    return true;
+}
+
+/// @brief Checks if there are any invalid buttons.
+/// @details Validatation with the previous input state doesn't happen because it doesn't exist.
+/// Therefore, we cannot check here if e.g. the drift button is pressed when it shouldn't be.
+/// @return If no invalid buttons are present.
+bool RaceInputState::isButtonsValid() const {
+    return !(buttons & ~0xf);
+}
+
+/// @brief Checks if the stick values are within the domain of the physics engine.
+/// @details The set of valid stick values is \f$\{\frac{x-7}{7}|0\leq x\leq 14,\in\mathbb{Z}\}\f$.
+/// It's possible for the stick input to be 8/7 with x = 15, but only with ghost controllers.
+/// @return If the stick values are valid.
+bool RaceInputState::isStickValid(f32 stick) const {
+    if (stick > 1.0f) {
```

Comment:
You don't have an equivalent check again `-1.0f` because the first loop iteration will catch this scenario. But so too will the "unreachable" return at the end of this function. I vote that you either add another early return check against `-1.0f` OR you get rid of this check and instead leverage the loop / end return

Reply:
Agreed, will add a check against `-1.0f`.